src/main/java/org/apache/spark/shuffle/ucx/memory/MemoryPool.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package org.apache.spark.shuffle.ucx.memory;
+
+import org.apache.spark.shuffle.UcxShuffleConf;
+import org.apache.spark.unsafe.Platform;
+import org.openucx.jucx.ucp.UcpContext;
+import org.openucx.jucx.ucp.UcpMemory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * Utility class to reuse and preallocate registered memory to avoid memory allocation
+ * and registration during shuffle phase.
+ */
+public class MemoryPool implements Closeable {
+  private static final Logger logger = LoggerFactory.getLogger(MemoryPool.class);
+
+  @Override
+  public void close() {
+    for (AllocatorStack stack: allocStackMap.values()) {
+      stack.close();
+      logger.info("Stack of size {}. " +
+          "Total requests: {}, total allocations: {}, preAllocations: {}",
+        stack.length, stack.totalRequests.get(), stack.totalAlloc.get(), stack.preAllocs.get());
+    }
+    allocStackMap.clear();
+  }
+
+  private class AllocatorStack implements Closeable {
+    private final AtomicInteger totalRequests = new AtomicInteger(0);
+    private final AtomicInteger totalAlloc = new AtomicInteger(0);
+    private final AtomicInteger preAllocs = new AtomicInteger(0);
+    private final ConcurrentLinkedDeque<RegisteredMemory> stack = new ConcurrentLinkedDeque<>();
+    private final int length;
+
+    private AllocatorStack(int length) {
+      this.length = length;
+    }
+
+    private RegisteredMemory get() {
+      RegisteredMemory result = stack.pollFirst();
+      if (result == null) {
+        ByteBuffer buffer = Platform.allocateDirectBuffer(length);
+        UcpMemory memory = null;
+        if (conf.preregisterMemory()) {
+          memory = context.registerMemory(buffer);
```

Comment:
so we register the receive buffer of every block separately?

Reply:
There are 2 branches: if no preallocation - then yes. If there's preallocation - we register contiguous buffer and slice it between buffers.

Reply:
why not have contig buffer for both cases? in case preallocation was not sufficient

Reply:
So 2 options now:

1. No preallocation: pretty simple. You request buffer of size N -> it allocates and returns you buffer. 
2. Preallocation: it preallocates K buffers of size N using 1 memory chunk of size: K*N

Reply:
i think it should be like UCX mpool: no (or small) preallocation, but gouws in large chunks

Reply:
It's an option number 1. But all the buffer allocation happens in the critical data path  (at the reducer). So we may preallocate needed buffers. That's why there's statistics there, you can run first without preallocation, get allocation statistics:
```
19/10/30 10:44:11,693 INFO MemoryPool: Stack of size 4096. Total requests: 205, total allocations: 25, preAllocations: 0
19/10/30 10:44:11,694 INFO MemoryPool: Stack of size 16777216. Total requests: 79, total allocations: 11, preAllocations: 0
19/10/30 10:44:11,694 INFO MemoryPool: Stack of size 8388608. Total requests: 111, total allocations: 15, preAllocations: 0
```
and after pre-allocate needed buffers